### PR TITLE
Auto-refresh of Tye services tree view

### LINKED ISSUE #2: Allow configuring of Tye service URI
We currently hardcode all interactions with tye to localhost:8000. This should be able to be changed.

### LINKED ISSUE #8: Tye tree view needs automatic refresh
There should be a mechanism to automatically refresh the Tye tree view, preferably upon receiving some signal to indicate a change, but at a minimum, on some periodic basis.  When in watch mode, the process IDs will cycle frequently, making it easy to try to attach the debugger to a long-gone process.

### LINKED ISSUE #69: Debug icon should be hidden when no application is running
Issue Type: <b>Bug</b>

When there is no application running, we should consider hiding the debug icon

image

Extension version: 0.1.0-alpha
VS Code version: Code 1.56.0 (cfa2e218100323074ac1948c885448fdf4de2a7f, 2021-05-04T22:09:06.405Z)
OS version: Windows_NT x64 10.0.19042

### PR DESCRIPTION
**Pull request by philliphoff**
Refactors the Tye application provider such that we can observe Tye applications as soon as they are started or shut down, to enable auto-refreshing of the Tye services tree view.  We now periodically scan the local system processes for Tye instances, then correlate them to open ports (to identify its dashboard port), which we can then use to query the application for its service metadata.

This also allows us to show multiple Tye applications in the tree view and, in concert with that, moves the "debug all" command to move from the tree view to each application node.  The UX when no Tye applications are running is also updated.

 
Screenshot 2021-06-25 at 10 51 11

Screenshot 2021-06-25 at 10 56 49

Resolves #2 (by making the issue moot)
Resolves #8
Resolves #69

### CODE REVIEW

File: src/services/portProvider.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+netstat.commands.darwin = {
+    cmd: 'lsof',
+    args: ['-Pn', '-i4', '-sTCP:LISTEN']
+};
+
+// NOTE: The TS definition incorrectly asserts that `darwin` is const.
+// TODO: Update the TS definitions.
+(netstat.parsers.darwin as unknown) = (line: string, callback: (item: netstat.ParsedItem) => void) => {
+    const parts = line.split(/\s/).filter(String);
+    if (!parts.length || (parts.length != 9 && parts.length != 10 )) {
```
Comment (pratiksanglikar) [inline src/services/portProvider.ts:15]: I'm confused about why parts of length 9 or 10 are an issue?
Maybe add a comment to elaborate?

Comment (philliphoff) [reply src/services/portProvider.ts:15]: I was following the pattern of the library's existing parsers; we just expect either 9 or 10 columns of data and bail if not found.  I've added a comment.

File: src/tasks/tyeRunTaskProvider.ts [hunk 2]
```diff
@@ -35,63 +33,55 @@ export interface TyeRunTaskDefinition extends TaskDefinition {
...
+                                        const applications = await tyeApplicationProvider.getApplications();                                               
+                                        const application = applications.find(application => application.pid === tyePid);
 
-                                            // NOTE: We arbitrarily pick the first application. This matches the tree view, which also shows only that first application.
-                                            //       Future work will refactor this logic to shutdown the appropriate application once Tye has better discovery support.
-                                            const application = applications[0];
+                                        if (application) {
                                             const tyeClient = tyeClientProvider(application.dashboard);
-
+
```
Comment (pratiksanglikar) [inline src/tasks/tyeRunTaskProvider.ts:84]: Nit: Stray whitespace

**Review by pratiksanglikar — APPROVED** [review]

File: src/tasks/tyeRunTaskProvider.ts (hunk 2 above)
Comment (philliphoff) [reply src/tasks/tyeRunTaskProvider.ts]: Removed.

**Pull request merged**